### ISSUE TITLE
TypeNotPresentException from getActualTypeArguments()

### ISSUE THREAD

**Issue by chadlowe3** [google/j2objc#940]
The following code is causing a TypeNotPresentException:

 ParameterizedType listTypeParams = (ParameterizedType) f.getGenericType();
 Type[] listTypeArgs = listTypeParams.**getActualTypeArguments()**;

f represents a List<QuoteBoardPage.PageEntry>

The stack trace is:

java.lang.TypeNotPresentException: Type com.t4login.application.quoteboard.settings.QuoteBoardPage$PageEntry not present
	at 0x00000001033e7cf0 java.lang.TypeNotPresentException.<init>() + 84
	at 0x000000010366f304 libcore.reflect.ParameterizedTypeImpl.getRawType() + 136
	at 0x000000010366f36c libcore.reflect.ParameterizedTypeImpl.getResolvedType() + 84
	at 0x000000010366e91c libcore.reflect.ListOfTypes.resolveTypes() + 264
	at 0x000000010366e7f4 libcore.reflect.ListOfTypes.getResolvedTypes() + 48
	at 0x000000010366f1e8 libcore.reflect.ParameterizedTypeImpl.getActualTypeArguments() + 36
	at 0x0000000103270d5c com.t4login.application.configuration.serialization.ConfigReader.parseConfigSettings() + 4784
	at 0x0000000103270964 com.t4login.application.configuration.serialization.ConfigReader.parseConfigSettings() + 3768
	at 0x00000001030645d0 com.t4login.application.quoteboard.settings.QuoteBoardLayout.restoreFromXML() + 476
	at 0x0000000102f8ecf8 com.t4login.api.usersettings.QuoteBoardLayoutParser.parse() + 144

I don't believe this is an issue with the reflection library. I created a test application and I was able to determine the generic type of a List<> with a nested and non-nested class just fine.

We just upgraded from J2ObjC version 1.0.2 and this same code worked fine with that version. So we are just wondering if you can help us figure out what we need to add to get this to work with version 2.0.5.

We have a prefixes.properties file and it is included in the copy bundle resources list. All of the other reflection code in the project works fine as far as we can tell.

We invoke J2ObjC with the following script:

"${J2OBJC_HOME}/j2objc" -d ${J2OBJC_DERIVED} -sourcepath "${T4_SOURCE_PATH}" --no-package-directories --swift-friendly --prefixes prefixes.properties -use-arc -g ${INPUT_FILE_PATH};

I am attaching the generated QuoteBoardPage.h file
[QuoteBoardPage.h.txt](https://github.com/google/j2objc/files/1686739/QuoteBoardPage.h.txt)

Any help would be greatly appreciated.

Thanks,
Chad

**Comment by tomball** [comment]
Thanks for the follow-up detail. I doubt this is related to type arguments or other reflection subtleties, as the stacktrace shows that Class.forName("com.t4login.application.quoteboard.settings.QuoteBoardPage$PageEntry") is failing and it should definitely work. That suggests two possible issues:

- There's a bug between how j2objc uses package prefixes and how the runtime does. What does your package.prefixes file look like? Do you have a specific entry for the `com.t4login.application.quoteboard.settings` package, or are you using wildcards? It shouldn't matter, but helps to create an accurate regression test.
- The QuoteBoardPage.o file is not being linked into your app. In your app's native code somewhere, does [NSClassFromString](https://developer.apple.com/documentation/foundation/1395135-nsclassfromstring?language=objc) return non-nil using the translated class name (prefix-from-prefixes-file + "QuoteBoardPage_PageEntry")? For these sorts of tests, I add them in my AppDelegate subclass's application: method, then either print the result using NSLog or set a breakpoint and run it in the debugger.

**Comment by tomball** [comment]
Regarding the NSClassFromString test, the class name should be what's declared in QuoteBoardPage.h for the PageEntry class.

**Comment by chadlowe3** [comment]
Our prefixes.properties file has:

```
com.t4login.application.chart.*: T4Chart
com.t4login.application.*: T4Application
com.t4login.*: T4
```

The NSClassFromString test only gave non-null when I omitted the prefix completely and used only the class name:

`    NSLog(@"QuoteBoardPage_PageEntry : %@", NSClassFromString(@"T4QuoteBoardPage_PageEntry"));`
2018-02-01 15:51:24.165888-0700 T4Mobile[951:3309907] T4Application.QuoteBoardPage_PageEntry : **T4QuoteBoardPage_PageEntry**

`    NSLog(@"T4Application.QuoteBoardPage_PageEntry : %@", NSClassFromString(@"T4Application.T4QuoteBoardPage_PageEntry"));`
2018-02-01 15:51:24.165937-0700 T4Mobile[951:3309907] T4Application.QuoteBoardPage_PageEntry : (null)

`    NSLog(@"T4Application.quoteboard.settings.T4QuoteBoardPage_PageEntry : %@", NSClassFromString(@"T4Application.quoteboard.settings.T4QuoteBoardPage_PageEntry"));`
2018-02-01 15:51:24.165948-0700 T4Mobile[951:3309907] T4Application.quoteboard.settings.T4QuoteBoardPage_PageEntry : (null)

`    NSLog(@"T4Application.QuoteBoardPage_PageEntry : %@", NSClassFromString(@"T4.application.quoteboard.settings.T4QuoteBoardPage_PageEntry"));`
2018-02-01 15:51:24.165963-0700 T4Mobile[951:3309907] T4.application.quoteboard.settings.T4QuoteBoardPage_PageEntry : (null)

`    NSLog(@"T4Application.QuoteBoardPage_PageEntry : %@", NSClassFromString(@"com.t4login.application.quoteboard.settings.T4QuoteBoardPage_PageEntry"));`
2018-02-01 15:51:24.165972-0700 T4Mobile[951:3309907] com.t4login.application.quoteboard.settings.T4QuoteBoardPage_PageEntry : (null)

**Comment by tomball** [comment]
Finding T4QuoteBoardPage_PageEntry shows QuoteBoardPage.o is linked in, so we can focus on why its prefix isn't mapping properly. In QuoteBoardPage.h, does it define T4QuoteBoardPage or T4ApplicationQuoteBoardPage (no dots)? Since that class can match on the second and third mapping lines, the translator and runtime may be iterating over lines in different orders. We're loading this file as a properties file, which extends HashTable, so it current iterates based on each key's hash value.

To address this, can I assume you expected matching to happen by checking each pattern in line order (line 1, then 2, then 3)? That seems to be "intuitively obvious" (nothing in API design actually is), so if that's your assumption, I can update both the translator and runtime to test patterns in line order.

**Comment by tomball** [comment]
This will take awhile to fix, since we can't break projects that have existing prefixes.properties files. In the meantime, you can avoid this issue by replacing the "com.t4login.*: T4" global wildcard with separate ones for each sub-package of com.t4login. This way, each package will only match one rule.

**Comment by chadlowe3** [comment]
That was it. I removed the T4Chart and T4Application prefixes and the TypeNotPresentException exception is no longer thrown.

The .h file defined the class as T4QuoteBoardPage. Poking around in other files, it appears that T4Chart was applied to a single class and T4Application was never actually applied.

I assumed that the prefixes were applied hierarchically. So everything under com.t4login.* would get T4, but since com.t4login.application.* is more specific, then everything under that would use T4Application instead of T4. Likewise everything under com.t4login.application.chart.* would use T4Chart as the path is more specific than T4Application.

It would be nice to break out the T4Chart namespace because of the sheer size of the codebase under it. However, we haven't started porting that part of the application over yet.

So we are actually resolved at this point.

Thank you very much for your help. We really appreciate the hard work your team has put into this valuable project!

**Comment by chadlowe3** [comment]
Replacing the global wildcard is a practical solution for us as well.

Thanks, again!

**Comment by antonio-cortes-perez** [comment]
This looks like duplicate of issue #995 which was fixed in version 2.2
